src/devices/src/virtio/test_utils.rs

```diff
@@ -329,3 +329,175 @@ impl<'a> VirtQueue<'a> {
+                virtqueues,
+            }
+        }
+
+        /// Get a (locked) reference to the device
+        pub fn device(&mut self) -> MutexGuard<T> {
+            self.device.lock().unwrap()
+        }
+
+        /// Activate the device
+        pub fn activate_device(&mut self, mem: &'a GuestMemoryMmap) {
+            self.device.lock().unwrap().activate(mem.clone()).unwrap();
+            // Process the activate event
+            let ev_count = self.event_manager.run_with_timeout(100).unwrap();
+            assert_eq!(ev_count, 1);
+        }
+
+        /// Get the start of the data region
+        ///
+        /// The first address that can be used for data in the guest memory mmap
+        /// is the first address after the memory occupied by the last Virtqueue
+        /// used by the device
+        pub fn data_address(&self) -> u64 {
+            self.virtqueues.last().unwrap().end().raw_value()
+        }
+
+        /// Add a new Descriptor in one of the device's queues
+        ///
+        /// This function adds in one of the queues of the device a DescriptorChain at some offset
+        /// in the "data range" of the guest memory. The number of descriptors to create is passed
+        /// as a list of descriptors (a triple of (index, length, flags)).
+        ///
+        /// The total size of the buffer is the sum of all lengths of this list of descriptors.
+        /// The fist descriptor will be stored at `self.data_address() + addr_offset`. Subsequent
+        /// descriptors will be placed at random addresses after that.
+        ///
+        /// # Arguments
+        ///
+        /// * `queue` - The index of the device queue to use
+        /// * `addr_offset` - Offset within the data region where to put the first descriptor
+        /// * `desc_list` - List of descriptors to create in the chain
+        pub fn add_desc_chain(
+            &mut self,
+            queue: usize,
+            addr_offset: u64,
+            desc_list: &[(u16, u32, u16)],
+        ) {
+            let device = self.device.lock().unwrap();
+
+            let event_fd = &device.queue_events()[queue];
+            let vq = &self.virtqueues[queue];
+
+            // Create the descriptor chain
+            let mut iter = desc_list.iter().peekable();
+            let mut addr = self.data_address() + addr_offset;
+            while let Some(&(index, len, flags)) = iter.next() {
+                let desc = &vq.dtable[index as usize];
+                desc.set(addr, len, flags, 0);
+                if let Some(&&(next_index, _, _)) = iter.peek() {
+                    desc.flags.set(flags | VIRTQ_DESC_F_NEXT);
+                    desc.next.set(next_index);
+                }
+
+                addr += u64::from(len);
+                // Add small random gaps between descriptor addresses in order to make sure we
+                // don't blindly read contiguous memory.
+                addr += u64::from(utils::rand::xor_pseudo_rng_u32()) % 10;
```

Comment:
Hmmm, I _think_ you are right, but we somehow need to be able to "emulate" the driver's behaviour. The spec allows buffers to not be in contiguous memory and it very often happens. So, I like the idea that are tests don't assume otherwise.

Reply:
Maybe we could seed a PRNG with some other value from the unit test?

Although I also saw the other day that this randomness is also present in other tests, so am happy to just address it outside of this PR